src/cache.rs

```diff
@@ -290,30 +290,57 @@ impl Cache {
 
     /// Parse a source and populate the corresponding entry in the cache, or do nothing if the
     /// entry has already been parsed.
-    pub fn parse(&mut self, file_id: FileId) -> Result<CacheOp<()>, ParseError> {
+    pub fn parse(&mut self, file_id: FileId) -> Result<CacheOp<()>, ParseErrors> {
         if self.terms.contains_key(&file_id) {
             Ok(CacheOp::Cached(()))
         } else {
-            self.terms
-                .insert(file_id, (self.parse_nocache(file_id)?, EntryState::Parsed));
+            self.terms.insert(
+                file_id,
+                (
+                    self.parse_nocache(file_id).map_err(|e| e.1)?,
+                    EntryState::Parsed,
+                ),
+            );
             Ok(CacheOp::Done(()))
         }
     }
 
+    /// Parse a source and populate the corresponding entry in the cache, or do nothing if the
+    /// entry has already been parsed.
+    ///
+    /// The cache will be populated even if there are parse errors.
+    pub fn parse_tolerate_errors(&mut self, file_id: FileId) -> Result<CacheOp<()>, ParseErrors> {
```

Comment:
The same remark as my other comment below (sorry, I did a non-linear review :sweat_smile: ): I dont think `ParseErrors` should go in the `Err` part of the result. IMO we should differentiate between recoverable parsing errors and fatal parsing errors in the return type.